File: nxflix-agents-ts/src/services/vocabulary.ts
import type { VocabularyItem, PartOfSpeech } from '../models/vocabulary.js';

export class VocabularyService {
  private vocabularyItems: Map<string, VocabularyItem> = new Map();
  private byWord: Map<string, string> = new Map();
  private byPartOfSpeech: Map<PartOfSpeech, string[]> = new Map();

  addVocabularyItem(vocab: VocabularyItem): void {
    this.vocabularyItems.set(vocab.id, vocab);
    this.byWord.set(vocab.word, vocab.id);

    const posList = this.byPartOfSpeech.get(vocab.partOfSpeech) ?? [];
    posList.push(vocab.id);
    this.byPartOfSpeech.set(vocab.partOfSpeech, posList);
  }

  getVocabularyItem(vocabId: string): VocabularyItem | undefined {
    return this.vocabularyItems.get(vocabId);
  }

  getVocabularyByWord(word: string): VocabularyItem | undefined {
    const id = this.byWord.get(word);
    return id ? this.vocabularyItems.get(id) : undefined;
  }

  getAllVocabularyItems(): VocabularyItem[] {
    return Array.from(this.vocabularyItems.values());
  }

  getVocabularyByPartOfSpeech(partOfSpeech: PartOfSpeech): VocabularyItem[] {
    const ids = this.byPartOfSpeech.get(partOfSpeech) ?? [];
    return ids
      .map((id) => this.vocabularyItems.get(id))
      .filter((v): v is VocabularyItem => v !== undefined);
  }

  getVocabularyItemsByIds(ids: string[]): VocabularyItem[] {
    return ids
      .map((id) => this.vocabularyItems.get(id))
      .filter((v): v is VocabularyItem => v !== undefined);
  }

  searchVocabulary(query: string): VocabularyItem[] {
    const lowerQuery = query.toLowerCase();
    const results: VocabularyItem[] = [];

    for (const vocab of this.vocabularyItems.values()) {
      if (
        vocab.word.includes(query) ||
        vocab.reading.includes(query) ||
        vocab.meanings.some((m) => m.toLowerCase().includes(lowerQuery))
      ) {
        results.push(vocab);
      }
    }

    return results;
  }

  get count(): number {
    return this.vocabularyItems.size;
  }

  get partsOfSpeech(): PartOfSpeech[] {
    return Array.from(this.byPartOfSpeech.keys());
  }
}

File: nxflix-agents-python/src/nxflix_agents/services/vocabulary.py
"""Vocabulary service for managing JLPT N1 vocabulary items."""

from nxflix_agents.models import VocabularyItem
from nxflix_agents.models.vocabulary import PartOfSpeech


class VocabularyService:
    """Service for managing vocabulary items in memory."""

    def __init__(self) -> None:
        self._vocabulary_items: dict[str, VocabularyItem] = {}
        self._by_word: dict[str, str] = {}
        self._by_part_of_speech: dict[PartOfSpeech, list[str]] = {}

    def add_vocabulary_item(self, vocab: VocabularyItem) -> None:
        """Add a vocabulary item to the service."""
        self._vocabulary_items[vocab.id] = vocab
        self._by_word[vocab.word] = vocab.id

        if vocab.part_of_speech not in self._by_part_of_speech:
            self._by_part_of_speech[vocab.part_of_speech] = []
        self._by_part_of_speech[vocab.part_of_speech].append(vocab.id)

    def get_vocabulary_item(self, vocab_id: str) -> VocabularyItem | None:
        """Get a vocabulary item by ID."""
        return self._vocabulary_items.get(vocab_id)

    def get_vocabulary_by_word(self, word: str) -> VocabularyItem | None:
        """Get a vocabulary item by its word."""
        vocab_id = self._by_word.get(word)
        return self._vocabulary_items.get(vocab_id) if vocab_id else None

    def get_all_vocabulary_items(self) -> list[VocabularyItem]:
        """Get all vocabulary items."""
        return list(self._vocabulary_items.values())

    def get_vocabulary_by_part_of_speech(
        self, part_of_speech: PartOfSpeech
    ) -> list[VocabularyItem]:
        """Get vocabulary items by part of speech."""
        ids = self._by_part_of_speech.get(part_of_speech, [])
        return [self._vocabulary_items[id_] for id_ in ids if id_ in self._vocabulary_items]

    def get_vocabulary_items_by_ids(self, ids: list[str]) -> list[VocabularyItem]:
        """Get multiple vocabulary items by their IDs."""
        return [
            self._vocabulary_items[id_] for id_ in ids if id_ in self._vocabulary_items
        ]

    def search_vocabulary(self, query: str) -> list[VocabularyItem]:
        """Search vocabulary by word, reading, or meanings."""
        lower_query = query.lower()
        results: list[VocabularyItem] = []

        for vocab in self._vocabulary_items.values():
            if (
                query in vocab.word
                or query in vocab.reading
                or any(m.lower().find(lower_query) != -1 for m in vocab.meanings)
            ):
                results.append(vocab)

        return results

    @property
    def count(self) -> int:
        """Get the number of vocabulary items."""
        return len(self._vocabulary_items)

    @property
    def parts_of_speech(self) -> list[PartOfSpeech]:
        """Get all available parts of speech."""
        return list(self._by_part_of_speech.keys())
